src/main/java/org/elasticsearch/index/translog/fs/FsTranslogFile.java

```diff
@@ -59,27 +74,85 @@ public static Type fromString(String type) throws ElasticsearchIllegalArgumentEx
         }
     }
 
-    long id();
 
-    int estimatedNumberOfOperations();
+    /** add the given bytes to the translog and return the location they were written at */
+    public abstract Translog.Location add(BytesReference data) throws IOException;
+
+    /** reuse resources from another translog file, which is guaranteed not to be used anymore */
+    public abstract void reuse(FsTranslogFile other) throws TranslogException;
+
+    /** change the size of the internal buffer if relevant */
+    public abstract void updateBufferSize(int bufferSize) throws TranslogException;
 
-    long translogSizeInBytes();
+    /** write all buffered ops to disk and fsync file */
+    public abstract void sync() throws IOException;
 
-    Translog.Location add(BytesReference data) throws IOException;
+    /** returns true if there are buffered ops */
+    public abstract boolean syncNeeded();
+
+    @Override
+    public FsChannelSnapshot snapshot() {
+        return new FsChannelSnapshot(immutableReader());
+    }
 
-    byte[] read(Translog.Location location) throws IOException;
+    /**
+     * returns a new reader that follows the current writes (most importantly allows making
+     * repeated snapshots that includes new content)
+     */
+    public FsChannelReader reader() {
+        channelReference.incRef();
+        boolean success = false;
+        try {
+            FsChannelReader reader = new InnerReader(this.id, channelReference);
+            success = true;
+            return reader;
+        } finally {
+            if (!success) {
+                channelReference.decRef();
+            }
+        }
+    }
 
-    FsChannelSnapshot snapshot() throws TranslogException;
 
-    void reuse(FsTranslogFile other) throws TranslogException;
+    /** returns a new immutable reader which only exposes the current written operation * */
+    abstract public FsChannelImmutableReader immutableReader();
 
-    void updateBufferSize(int bufferSize) throws TranslogException;
+    boolean assertBytesAtLocation(Translog.Location location, BytesReference expectedBytes) throws IOException {
+        ByteBuffer buffer = ByteBuffer.allocate(location.size);
+        readBytes(buffer, location.translogLocation);
+        return new BytesArray(buffer.array()).equals(expectedBytes);
+    }
 
-    void sync() throws IOException;
+    /**
+     * this class is used when one wants a reference to this file which exposes all recently written operation.
+     * as such it needs access to the internals of the current reader
+     */
+    class InnerReader extends FsChannelReader {
```

Comment:
it just delegates to the other reader for enforcement. We need it to maintain the sematics of using a reader only in one place and all reference counting is done on the channel.